SDK Revamp

br41nslug opened the pull request:
The design goals for this rebuild:
- TypeScript first
- Modular/Composable architecture
- Lightweight and Dependency Free

## Features

- Authentication composables for automatic refreshing and static tokens
- REST feature parity with the API
- Basic GraphQL request helper
- RealTime Websocket Client

## Examples

### Client
```ts
const client = createDirectus<Schema>('https://api.directus.io')
    .with(rest())
    .with(graphql());

// do a REST request
const restResult = await client.request(readItems('articles'));

// do a GraphQL request
const gqlResult = await client.query<OutputType>(`
    query {
        articles {
            id
            title
            author {
                first_name
            }
        }
    }
`);
```

### Authentication
```ts
const client = createDirectus<Schema>('https://api.directus.io')
    .with(rest())
    .with(authentication('json'));

await client.login('admin@example.com', 'd1r3ctu5');

// do authenticated requests
```
```ts
const client = createDirectus<Schema>('https://api.directus.io')
    .with(rest())
    .with(staticToken('super-secure-token'));

// do authenticated requests
```

### Real-Time
```ts
const client = createDirectus<Schema>('https://api.directus.io')
    .with(realtime({ authMode: 'public' }));

const { subscription, unsubscribe } = await client.subscribe('test', {
	  query: { fields: ['*'] },
});

for await (const item of subscription) {
	  console.log('subscription', { item });
}

// unsubscribe()
```

```ts
const client = createDirectus<Schema>('https://api.directus.io').with(
	realtime({
		authMode: 'public',
	})
);

const stop = client.onWebSocket('message', (message) => {
	if ('type' in message && message['type'] === 'pong') {
		console.log('PONG received');
		stop();
	}
});

client.sendMessage({ type: 'ping' });
```

### Build your schema

```ts
// The main schema type containing all collections available
interface MySchema {
	collection_a: CollectionA[]; // regular collections are array types
	collection_b: CollectionB[];
	collection_c: CollectionC; // this is a singleton
	// junction collections are collections too
	collection_a_b_m2m: CollectionAB_Many[];
	collection_a_b_m2a: CollectionAB_Any[];
}

// collection A
interface CollectionA {
	id: number;
	status: string;
	// relations
	m2o: number | CollectionB;
	o2m: number[] | CollectionB[];
	m2m: number[] | CollectionAB_Many[];
	m2a: number[] | CollectionAB_Any[];
}

// Many-to-Many junction table
interface CollectionAB_Many {
	id: number;
	collection_a_id: CollectionA;
	collection_b_id: CollectionB;
}

// Many-to-Any junction table
interface CollectionAB_Any {
	id: number;
	collection_a_id: CollectionA;
	collection: 'collection_b' | 'collection_c';
	item: string | CollectionB | CollectionC;
}

// collection B
interface CollectionB {
	id: number;
	value: string;
}

// singleton collection
interface CollectionC {
	id: number;
	app_settings: string;
	something: string;
}
```

## TODO
- [x] REST Composable
  - [ ] Improve advanced query option types (alias, deep)
  - [ ] Error handling hooks
- [ ] GraphQL Subscriptions Composable
- [x] Real-Time Composable
  - [ ] Improve error handling
  - [ ] Implement CRUD message commands
- [ ] Pagination Composable
- [ ] Unit Tests

rijkvanzanten committed:
Setup build chain

rijkvanzanten committed:
Initial structure

rijkvanzanten committed:
Typing, exports, fix build

rijkvanzanten committed:
Reorganize

br41nslug committed:
prutsen met output types

rijkvanzanten committed:
Thursday fun times

br41nslug committed:
failed experiments in feature composition

br41nslug committed:
got feature flags on the client

br41nslug committed:
using interfaces instead of literals

br41nslug committed:
messed with decorators

br41nslug committed:
split up decorators

br41nslug committed:
added fetch and ws types

br41nslug committed:
reintroduced the global.d.ts we'll figure that out later

br41nslug committed:
reworked composables

br41nslug committed:
some changes

br41nslug committed:
experimenting with commands

br41nslug committed:
got the read item command working again with types

br41nslug committed:
trying authentication

br41nslug committed:
basic gql implementation

br41nslug committed:
Super basic websocket implementation

br41nslug committed:
renamed websocket composable and removed obsolete code

br41nslug committed:
Typing relational fields

Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

br41nslug committed:
fixed recursive type

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

rijkvanzanten committed:
SDK revamp revamp (#18916)

* Split up files

* Simplify some more (too much?)

* Oh boy we figured out object extending

* OK JK now we're there

* OK getting somewhere real now

* Start structuring type helpers

* Types werken zowaar!

* Remove unneeded "T"'s for readability

* Added support for array relational types

* split up query nesting logic

* make relational fields optional to support multiple relations

* improved readability

* updated REST

* Move URL to top level

* Remove T

* recursively join fields in params

* cleaning up

* updated graphql implementation

* removing old unused code

---------

Co-authored-by: Brainslug <tim@brainslug.nl>

rijkvanzanten committed:
Reduce global reuse of rest

rijkvanzanten committed:
Start messing with output types

rijkvanzanten committed:
Return it from the rest command

br41nslug committed:
first level schema output filtering

br41nslug committed:
It works!! but will need some docs for context

br41nslug committed:
Added some comments

br41nslug committed:
removed export keyword from types that do not need to be exported

br41nslug committed:
cleaned up debug script

br41nslug committed:
graphql update

br41nslug committed:
Added dom for WebSocket and Fetch typings

br41nslug committed:
initial readme setup

br41nslug committed:
moved some types

br41nslug committed:
removed accidental commit

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

fabian-hiller committed:
Refactor `useDirectus` and improve token handling (#18966)

* Add comment and reduce code of useDirectus

* Add fabian-hiller to contributors.yml

* Change token handling of useDirectus client

---------

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

fabian-hiller committed:
Change ClientConfig type back to interface (#18968)

br41nslug committed:
initial realtime feature

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
started on some ws commands and implemented message receiving callback

br41nslug committed:
disabled rest config for now

br41nslug committed:
setup auth extension boilerplate

br41nslug committed:
Added some websocket examples to the readme

estheragbaje committed:
docs: add getting started guide

estheragbaje committed:
docs: add link to previous sdk

estheragbaje committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

estheragbaje committed:
docs: updates

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
implemented readItem

paescuj commented on docs/reference/sdk.md:
```diff
@@ -12,6 +12,12 @@ readTime: 14 min read
+<Card
+  title="Explore the new JavaScript SDK"
+  h="2"
+  text="The new SDK is a significant overhaul of the previous version."
```
Or even complete re-implementation? 🤔

br41nslug committed:
updated readItems

br41nslug committed:
Added delete commands

estheragbaje committed:
docs: updates to snippets and phrasing

estheragbaje committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
updated lockfile

paescuj committed:
Update dictionary

estheragbaje committed:
docs: removed repeated typescript snippet

estheragbaje committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

paescuj committed:
Enrich package.json for SDK

estheragbaje committed:
docs: update snippet

estheragbaje committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

paescuj committed:
Fix casing of readme

br41nslug committed:
added create items commands

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
fixed read items output

br41nslug committed:
removed obsolete types

br41nslug committed:
added update commands

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

paescuj commented:
Currently, types aren't available for CJS:
```
❯ attw directus-sdk-11.0.0.tgz

💀 Import failed to resolve to type declarations or JavaScript files. https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/NoResolution.md

❌ Import resolved to JavaScript files, but no type declarations were found. https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/UntypedResolution.md


┌────────────────────┬───────────────────────────────────┐
│                    │ "@directus/sdk"                   │
├────────────────────┼───────────────────────────────────┤
│ node10             │ 💀 Resolution failed              │
├────────────────────┼───────────────────────────────────┤
│ node16 (from CJS)  │ ❌ No types                       │
├────────────────────┼───────────────────────────────────┤
│ node16 (from ESM)  │ 🟢 (ESM)                          │
├────────────────────┼───────────────────────────────────┤
│ bundler            │ 🟢                                │
└────────────────────┴───────────────────────────────────┘
```

Could be easily solved with https://github.com/egoist/tsup/pull/934 being merged and released, but it looks like we've to come up with a workaround in the meantime.

br41nslug committed:
updated tsup

br41nslug commented:
Updating `tsup` to 7.1 which includes that PR did the trick.
![firefox_02yHRtaOO8](https://github.com/directus/directus/assets/9389634/87f16734-a5dc-4d18-a563-c83e4394eb4d)

paescuj commented:
> Updating `tsup` to 7.1 which includes that PR did the trick. ![firefox_02yHRtaOO8](https://user-images.githubusercontent.com/9389634/248768381-87f16734-a5dc-4d18-a563-c83e4394eb4d.png)

Wow, awesome! Didn't expect it to get merged & released that fast 🚀❤️😄

br41nslug committed:
removed unused dependencies

br41nslug committed:
implemented basic memory storage fallback

br41nslug committed:
implemented the login function for auth

br41nslug committed:
implemented refresh and logout auth functions

estheragbaje committed:
docs: update guide

andrewbranch commented:
Me either! 😄 So cool to see it helping out with a real example already!

br41nslug committed:
updated auth refresh logic

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
oops, have to actually store the expires at value

br41nslug committed:
fixed authentication timeout

br41nslug committed:
added autoRefresh config option

phazonoverload commented on docs/guides/sdk/getting-started.md:
```diff
@@ -0,0 +1,226 @@
+
+::: warning SDK in Beta
+
+The new SDK is in Beta, so we recommend using it in a staging environment first before deploying it to production.
```
```suggestion
The new SDK is in Beta, so we do not recommend using it in production.
```

phazonoverload commented on docs/guides/sdk/getting-started.md:
```diff
@@ -0,0 +1,226 @@
+
+- `rest()`: REST request functions, adds `.request(...)` to the client.
+- `graphql()`: GraphQL request functions, adds `.query(...)` to the client.
+- `auth()`: Authentication functions, adds `.login(...)`, `.logout()`, and .`refresh()` to the client. (work in
+  progress).
+- `realtime()`: WebSocket connectivity, adds `.subscribe(...)` and `.message(...)` to the client. (work in progress).
+- `subscription()`: GraphQL Subscriptions, adds `.subscription()` to the client. (work in progress).
```
@estheragbaje @br41nslug can we add a note on what 'Work in progress' means here? Does it work... at all? Is it indicative of coming soon functionality? Need to set expectations clearly.

br41nslug replied:
The parts marked work in progress are planned but either still under development or todo and very likely dont work yet. i have updated the readme and getting started.

phazonoverload commented on docs/guides/sdk/getting-started.md:
```diff
@@ -0,0 +1,226 @@
+If you encounter any issues or have questions, don't hesitate to reach out to us in our
+[Discord community](https://directus.chat/).
+
+## Backward Compatibility
+
+The older Directus SDK will still work and your existing applications that use the older version will continue to
+function without any issues.
```
This feels wrong. I'm reading it as "this is not backwards compatible" (because it isn't) "but feel free to keep using the older versions".

estheragbaje replied:
I changed the heading to "Older SDK Version". I think that makes the content more understandable

phazonoverload commented on docs/guides/sdk/getting-started.md:
```diff
@@ -0,0 +1,226 @@
+import { useDirectus } from '@directus/sdk@beta';
+import { auth } from '@directus/sdk@beta/composable';
+
+const client = useDirectus('http://directus.example.co').use(auth());
```
```suggestion
const client = useDirectus('http://directus.example.com').use(auth());
```

phazonoverload reviewed (commented):
Some small changes on the getting started guide. Looks good, reads well. I also think we should link to the Typedoc (where will that be, @rijkvanzanten?) so folks can discover the composables themselves.

br41nslug committed:
updated readme

br41nslug committed:
updated getting started

estheragbaje committed:
docs: updates to guide

br41nslug committed:
added beta version nr

br41nslug committed:
removed debug scripts

br41nslug committed:
added docblocks to the composables

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
Added some docblocks

br41nslug committed:
Added BETA warning to readme

br41nslug committed:
Added extra type extends object checks

br41nslug committed:
added a bunch of index files

br41nslug committed:
updated tsup build entrypoints and exports

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
updated import paths

br41nslug committed:
updated code examples

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug commented on docs/guides/sdk/getting-started.md:
````diff
@@ -0,0 +1,250 @@
+`);
+```
+
+> Feel free to customize the queries according to your project's GraphQL schema.
````
Does this add anything to the example? any code example should be customized/adapted for your own project if copied.

phazonoverload replied:
What do you mean?

br41nslug replied:
the comment is only about the quote itself

br41nslug commented on docs/guides/sdk/getting-started.md:
```diff
@@ -0,0 +1,250 @@
+If you encounter any issues or have questions, don't hesitate to reach out to us in our
+[Discord community](https://directus.chat/).
+
+## Older SDK Version
+
+The older SDK still works and your existing applications that use the older version will continue to
+function without any issues.
```
Should we maybe just drop this? It feels somewhat misleading as `"existing applications that use the older version will continue to function without any issues."` only holds true if they have pinned the dependency to their old version

estheragbaje replied:
That makes sense, I'll take it out.

estheragbaje committed:
docs: removed unnecessary phrases

paescuj committed:
Unignore SDK changesets

br41nslug committed:
Revert "docs: removed unnecessary phrases"

This reverts commit 3559ade87316782701a5c0dbe809e3c3864ac4cd.

br41nslug committed:
docs: removed unnecessary phrases

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

rijkvanzanten committed:
Make sure we export all available functions

rijkvanzanten committed:
Tweak exports

rijkvanzanten committed:
Re-add beta version flag in package.json

paescuj committed:
Add TypeDoc for new (and old) SDK

paescuj committed:
Merge branch 'main' into sdk-revamp

paescuj commented on docs/guides/sdk/getting-started.md:
````diff
@@ -0,0 +1,243 @@
+```ts
+import { useDirectus } from '@directus/sdk@beta';
+
+// your Directus collection schema
````
```suggestion
// Your Directus collection schema
```

paescuj committed:
Update docs/guides/sdk/getting-started.md

rijkvanzanten committed:
Format readme

rijkvanzanten committed:
Tweaks

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

rijkvanzanten committed:
v10.4.0

paescuj committed:
Spellchecker

paescuj committed:
Add link to TypeDoc of new SDK

br41nslug committed:
updated query types

br41nslug committed:
trying to expand the query types

br41nslug committed:
extending the Query options

br41nslug committed:
add all exports to the root

br41nslug committed:
fixed incorrect field mapping

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
Abstracted more  of the request logic for more flexibility

br41nslug committed:
partial error handling, global fetch settings and fetch response options

br41nslug committed:
reworked the requests for better options control and started on better error handling

br41nslug committed:
type fix

paescuj committed:
Merge branch 'main' into sdk-revamp

paescuj committed:
Merge branch 'main' into sdk-revamp

paescuj committed:
Run formatter & linter

paescuj committed:
Spell fixes

br41nslug committed:
added onRequest handler

br41nslug committed:
updated global request hooks

br41nslug committed:
expanded ItemType to accept singletons

br41nslug committed:
updated output types for singletons

br41nslug committed:
fixed fields type generation

br41nslug committed:
removed rest globaloptions in favor of onRequest

br41nslug committed:
started singleton command

br41nslug committed:
different commands for singletons

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
ran prettier

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
updated readme

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
stateless helpers for tomorrow

br41nslug committed:
support m2a in field types

br41nslug committed:
added m2a fields parsing and attempt at output typing

br41nslug committed:
made the base client stateless

br41nslug committed:
making sure things are exported as needed

br41nslug committed:
added minimal core table schema for files and users

br41nslug committed:
added update singleton command

br41nslug committed:
Added core user read commands

br41nslug committed:
merged related commands into the same file

br41nslug committed:
Added create users commands

br41nslug committed:
Added user update commands

br41nslug committed:
added user delete commands

br41nslug committed:
updated query handling

br41nslug committed:
ran prettier and bumped version

br41nslug committed:
ran prettier

br41nslug committed:
disabled global error handling for now

br41nslug committed:
fixed output formatting for delete commands

br41nslug committed:
fixed onResponse handling

br41nslug committed:
fixed item types for create and update

br41nslug committed:
added missing partial item for create

br41nslug committed:
ran prettier

br41nslug committed:
updated readme

br41nslug committed:
added missing export

rijkvanzanten committed:
Publish beta version

br41nslug committed:
Added a bunch of core collection types

br41nslug committed:
updated users commands

br41nslug committed:
added more complex core types

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
double checked nullable core fields

br41nslug committed:
fixed core collection merging with custom properties

br41nslug committed:
added authentication commands

br41nslug committed:
ran prettier and fixed line endings

br41nslug committed:
Added all core read commands

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
Added all core create commands

br41nslug committed:
Added all core delete commands

br41nslug committed:
Added all core update commands

br41nslug committed:
ran prettier

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
renamed `useDirectus` to `createDirectus` and `.use` to `.with` to prevent naming confusion with reacthooks or express middleware.

br41nslug committed:
ran prettier

br41nslug committed:
added more misc core commands

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
added schema endpoints

br41nslug committed:
added server api commands

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
ran prettier

br41nslug committed:
added most utility endpoints

br41nslug committed:
finished utility endpoints

br41nslug committed:
checking the command typing and jsdoc

br41nslug committed:
fixed those pesky relational output types

br41nslug committed:
ran prettier

br41nslug committed:
removed obsolete type

br41nslug committed:
initial aggregation typing

br41nslug committed:
improved aggregation and grouping typing

br41nslug committed:
ran prettier

br41nslug committed:
checked exports

br41nslug committed:
Merge branch 'main' into sdk-revamp

rijkvanzanten committed:
Beta 4

br41nslug committed:
fixed the m2a fields regression

br41nslug committed:
added basic client test

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
do query parsing at the latest stage

br41nslug committed:
extra type comments

br41nslug committed:
reverted accidental commit and ran prettier

br41nslug committed:
Merge branch 'main' into sdk-revamp

paescuj committed:
Adjust tsconfig to updated @directus/tsconfig package

br41nslug committed:
updated websocket event handling

br41nslug committed:
fixed capitalization

br41nslug committed:
removed obsolete queryToParams usage

br41nslug committed:
Merge branch 'main' into sdk-revamp

br41nslug committed:
implemented websocket reconnecting and some non-happy path fixes

br41nslug committed:
persist subscription across reconnecting

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
updated readme

br41nslug committed:
ran prettier

rijkvanzanten committed:
11.0.0-beta.5

br41nslug committed:
using plurar for uploadFiles ad removed "D" from updateFiles

br41nslug committed:
fix incorrect paths

br41nslug committed:
created new util types file

br41nslug committed:
extracted output types

br41nslug committed:
fixed input typing issue (for items)

br41nslug committed:
extracted functions to its own file

br41nslug committed:
fixed path for presets

br41nslug committed:
ran prettier

br41nslug committed:
Applied typing fix to all commands

br41nslug committed:
fixed naming error `updatedItems` -> `updateItems`

br41nslug committed:
fixed type error

br41nslug committed:
Failing types gracefully if no schema is provided

paescuj commented on sdk/src/schema/preset.ts:
```diff
@@ -0,0 +1,23 @@
+	{
+		id: number;
+		bookmark: string | null;
+		user: DirectusUser<Schema> | string | null;
+		role: DirectusRole<Schema> | string | null;
```
Should be just as it only accepts / returns the UUIDs
```suggestion
		user: string | null;
		role: string | null;
```
or maybe
```suggestion
		user: DirectusUser<Schema>['id'] | null;
		role: DirectusRole<Schema>['id'] | null;
```

br41nslug replied:
I dont think that is true 🤔 the `/presets` endpoint does accept the `fields` parameter so `?fields=*,user.*` will fetch you a complete user object

paescuj replied:
Oh, I see 👍 But not for the request, right?

paescuj replied:
Discussed on Discord, types are correct ✔️

rijkvanzanten committed:
Merge branch 'main' into sdk-revamp

rijkvanzanten committed:
beta.6

paescuj commented on sdk/src/schema/webhook.ts:
```diff
@@ -0,0 +1,17 @@
+		url: string;
+		status: string;
+		data: boolean;
+		actions: string;
+		collections: string;
```
Fields of type `csv` always seem to return an array not a string while accepting both, an array and string for requests. At least confirmed in this case for `createWebhook`:
```js
// requests
actions: 'create'
actions: ['create']
actions: 'create,delete'
actions: ['create', 'delete']

// responses
actions: [ 'create' ]
actions: [ 'create' ]
actions: [ 'create', 'delete' ]
actions: [ 'create', 'delete' ]
```

This means the response type is clearly wrong. However, I'm not sure how we can best deal with something like this, since request / response types are different 🤔

paescuj committed:
Clean-up tsup config

- Use documented way to define config
- Doesn't need to be ts file: not type checked, intellisense available
  in vscode anyway
- remove default & deprecated options

br41nslug committed:
added other user commands

br41nslug committed:
added other share commands

br41nslug committed:
Merge branch 'sdk-revamp' of https://github.com/directus/directus into sdk-revamp

br41nslug committed:
satisfy prettier

br41nslug committed:
added string array types to webhook core schema

br41nslug committed:
fixed plain string request output

br41nslug committed:
keep nullability for relational fields

br41nslug committed:
Merge branch 'main' into sdk-revamp

estheragbaje committed:
Update documentation guides to use new SDK (#19131)

Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>
Co-authored-by: Kevin Lewis <kvn@lws.io>
Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>
Co-authored-by: Brainslug <tim@brainslug.nl>

paescuj committed:
Merge branch 'main' into sdk-revamp

paescuj committed:
Fix error messages for core collections

br41nslug committed:
updated dictionary

br41nslug committed:
updated partial item input for fields

paescuj committed:
Ready, set, go!
